### PR TITLE
feat(router): implement API endpoints for managing API keys

### LINKED ISSUE #501: [FEATURE] Allow rotating and revoking API keys
### Feature Description

As of now, the `merchant_account` table has an `api_key` field which restricts a merchant account to only have one API key. Furthermore, there are no mechanisms in place to revoke and regenerate API keys. To improve the convenience of rotating and revoking API keys, it'd be better to allow multiple API keys to be associated with a merchant account. Furthermore, API keys are stored in plaintext as of now. It'd be great if they could be hashed as well, when this request is addressed.

### Possible Implementation

A possible approach to address this is to add an `api_keys` table with `name`, `merchant_id`, `hashed_key` fields (including any additional fields as necessary). And for the API interface, this could include endpoints to create, retrieve, update and revoke an API key, and an endpoint to list all API keys associated with the current merchant account.

### Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### Have you read the Contributing Guidelines?

- [X] I have read the [Contributing Guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md)

### Are you willing to submit a PR?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by SanchithHegde**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description

This PR implements the endpoints to create, retrieve, update, revoke and list API keys. 

- As of now, the `last_used` field is included in the storage model, but I don't expect it to be actively used until API keys are cached in Redis or in-memory.
- The endpoints use admin API key authentication, but require an `X-Merchant-Id` header to be sent for creating an API key and to list API keys for a specific merchant.

This PR only implements the endpoints for managing API keys, I'll update merchant authentication to use these API keys in a separate PR.

### Additional Changes

- [x] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

The relevant migrations are [`up.sql`](https://github.com/juspay/hyperswitch/blob/api-keys-endpoints/migrations/2023-02-01-135102_create_api_keys_table/up.sql) and [`down.sql`](https://github.com/juspay/hyperswitch/blob/api-keys-endpoints/migrations/2023-02-01-135102_create_api_keys_table/down.sql).

## Motivation and Context

Closes #501.

## How did you test it?

I have added a few unit tests where database accesses are not involved.

I'll attach invocations of the APIs via `curl` to give an idea of how they look.

- Create API Key:
  
  ```shell
  curl --location --request POST 'http://localhost:8080/api_keys' \
  --header 'Content-Type: application/json' \
  --header 'Accept: application/json' \
  --header 'X-Merchant-Id: merchant123' \
  --header 'api-key: MyVerySecretAdminApiKey' \
  --data-raw '{
  "name": "API Key 1",
  "description": null,
  "expiration": "2023-09-23T01:02:03.000Z"
  }'
  {
    "key_id": "dev_PwX5u10PjAbs9XWac0B2",
    "merchant_id": "merchant123",
    "name": "API Key 1",
    "description": null,
    "api_key": "dev_PwX5u10PjAbs9XWac0B2-MyVerySecret64CharacterLongApiKey",
    "created": "2023-02-06T14:16:46.144Z",
    "expiration": "2023-09-23T01:02:03.000Z"
  }
  ```

- Retrieve API Key:

  ```shell
  curl --location --request GET 'http://localhost:8080/api_keys/dev_PwX5u10PjAbs9XWac0B2' \
  --header 'Content-Type: application/json' \
  --header 'Accept: application/json' \
  --header 'api-key: MyVerySecretAdminApiKey'
  {
    "key_id": "dev_PwX5u10PjAbs9XWac0B2",
    "merchant_id": "merchant123",
    "name": "API Key 1",
    "description": null,
    "prefix": "dev_PwX5u10PjAbs9XWac0B2-MyVerySe",
    "created": "2023-02-06T14:16:46.144Z",
    "expiration": "2023-09-23T01:02:03.000Z"
  }
  ```

- Update API Key:

  ```shell
  curl --location --request POST 'http://localhost:8080/api_keys/dev_PwX5u10PjAbs9XWac0B2' \
  --header 'Content-Type: application/json' \
  --header 'Accept: application/json' \
  --header 'api-key: MyVerySecretAdminApiKey' \
  --data-raw '{
    "description": "My very awesome API key"
  }'
  {
    "key_id": "dev_PwX5u10PjAbs9XWac0B2",
    "merchant_id": "merchant123",
    "name": "API Key 1",
    "description": "My very awesome API key",
    "prefix": "dev_PwX5u10PjAbs9XWac0B2-MyVerySe",
    "created": "2023-02-06T14:16:46.144Z",
    "expiration": "2023-09-23T01:02:03.000Z"
  }
  ```

- Revoke API Key:

  ```shell
  curl --location --request DELETE 'http://localhost:8080/api_keys/dev_PwX5u10PjAbs9XWac0B2' \
  --header 'Content-Type: application/json' \
  --header 'Accept: application/json' \
  --header 'api-key: MyVerySecretAdminApiKey' \
  --data-raw ''
  {
    "key_id": "dev_PwX5u10PjAbs9XWac0B2",
    "revoked": true
  }
  ```

- List API Keys

  ```shell
  curl --location --request GET 'http://localhost:8080/api_keys/list' \
  --header 'Content-Type: application/json' \
  --header 'Accept: application/json' \
  --header 'X-Merchant-Id: merchant123' \
  --header 'api-key: MyVerySecretAdminApiKey' \
  --data-raw ''
  [
    {
      "key_id": "dev_PZ7uBeZrBmBrhXmLo3W0",
      "merchant_id": "merchant123",
      "name": "API Key 1",
      "description": null,
      "prefix": "dev_PZ7uBeZrBmBrhXmLo3W0-ApiKey01",
      "created": "2023-02-05T18:51:04.903Z",
      "expiration": "2023-09-23T01:02:03.000Z"
    },
    {
      "key_id": "dev_gd6zpyE2GDnjqVGyYGVN",
      "merchant_id": "merchant123",
      "name": "API Key 2",
      "description": null,
      "prefix": "dev_gd6zpyE2GDnjqVGyYGVN-ApiKey02",
      "created": "2023-02-05T18:51:08.272Z",
      "expiration": "2023-09-23T01:02:03.000Z"
    },
    {
      "key_id": "dev_8RPcI7zCs3c37KjsGYT0",
      "merchant_id": "merchant123",
      "name": "API Key 3",
      "description": null,
      "prefix": "dev_8RPcI7zCs3c37KjsGYT0-ApiKey03",
      "created": "2023-02-05T18:51:11.651Z",
      "expiration": "2023-09-23T01:02:03.000Z"
    }
  ]
  ```
## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/routes/app.rs [hunk 1]
```diff
@@ -329,3 +331,21 @@ impl Configs {
             )
     }
 }
+
+pub struct ApiKeys;
+
+#[cfg(any(feature = "olap", feature = "oltp"))]
```
Comment (jarnura) [inline crates/router/src/routes/app.rs:337]: All routes won't be for both `olap` and `oltp`.

All Dashboard using will be olap and only the backend api comes in oltp.

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/api_keys.rs [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+use crate::{
+    consts,
+    core::errors::{self, RouterResponse, StorageErrorExt},
+    db::StorageInterface,
+    services::ApplicationResponse,
+    types::{api, storage, transformers::ForeignInto},
+    utils,
+};
+
+// Defining new types `PlaintextApiKey` and `HashedApiKey` in the hopes of reducing the possibility
```
Comment (ashokkjag) [inline crates/router/src/core/api_keys.rs:15]: Good to enforce this using types. 👍

File: crates/router/src/core/api_keys.rs [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+// of plaintext API key being stored in the data store.
+pub struct PlaintextApiKey(Secret<String>);
+pub struct HashedApiKey(String);
+
+impl PlaintextApiKey {
+    const HASH_KEY_LEN: usize = 32;
+
+    const PREFIX_LEN: usize = 12;
+
+    pub fn new(length: usize) -> Self {
```
Comment (ashokkjag) [inline crates/router/src/core/api_keys.rs:25]: is key generation based on the environment in which its running? Or should this be pased? Or should we always generate a snd / prd key pair?

**Comment by ashokkjag** [comment]
LGTM, but few questions. 
- internalservererror for few of the routes. should that be empty 200 ok instead? 
- will the key-id be another HTTP header?

File: crates/router/src/core/api_keys.rs (hunk 3 above)
Comment (SanchithHegde) [reply crates/router/src/core/api_keys.rs:25]: It is based on the environment in which the app is running. The information is obtained from the `router_env` crate:

https://github.com/juspay/hyperswitch/blob/8205b2278d5a745d047274f17798a55e4c790a7f/crates/router_env/src/env.rs#L31-L38

**Comment by SanchithHegde** [comment]
@ashokkjag 

> * internalservererror for few of the routes. should that be empty 200 ok instead?

I've updated the PR to raise a 400 status code in case an invalid `X-Merchant-Id` header value is provided. I checked the other places where 500 is returned, I couldn't find any instances where we could return a 200 instead.

> * will the key-id be another HTTP header?

Yes, that change (to use these API keys for merchant authentication) will be taken up on a separate PR.

**Review by jarnura — DISMISSED** [review]

**Comment by SanchithHegde** [comment]
Marking this `S-in-progress` since the format in which plaintext API keys are returned might change slightly.

File: crates/router/src/core/api_keys.rs [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+        let env = match router_env::env::which() {
+            router_env::Env::Development => "dev",
+            router_env::Env::Sandbox => "snd",
+            router_env::Env::Production => "prd",
```
Comment (NishantJoshi00) [inline crates/router/src/core/api_keys.rs:29]: Hey, can we define this somewhere in the `router_env` itself, we have a similar match case in `admin.rs`. have it in a single place will help in the future, if a need of performing a similar match arises 

https://github.com/juspay/hyperswitch/blob/3292960996834f93095f96f05b18ce28adceb374/crates/router/src/core/admin.rs#L23-L27

**Review by NishantJoshi00 — COMMENTED** [review]
Other than that, looks good to me!

**Review by ashokkjag — APPROVED** [review]

**Comment by SanchithHegde** [comment]
Sorry I forgot to send an update here, I've updated the PR so that the plaintext API key returned on creation is of the format: `{key_id}-{api_key}`. With this change, the merchant would have to send only one value as before, and we can handle merchant authentication without much effort. 

To go with this change, the key ID now has the prefix based on the environment, one of `dev`, `snd` and `prd` instead of the `key` prefix used in my previous implementation.

**Review by jarnura — APPROVED** [review]

**Pull request merged**